lib/new_relic/agent/database_adapter.rb

```diff
@@ -0,0 +1,73 @@
+# encoding: utf-8
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+
+module NewRelic
+  module Agent
+    class DatabaseAdapter
+      class Base
+        attr_reader :env, :version
+
+        def initialize(env, version)
+          @env = env
+          @version = Gem::Version.new(version)
+        end
+
+        def match?
+          true
+        end
+
+        def adapter
+          nil
+        end
+      end
+
+      class Rails30 < Base
+        def match?
+          version >= Gem::Version.new('3.0.0')
+        end
+
+        def adapter
+          ActiveRecord::Base.configurations[env]['adapter']
+        end
+      end
+
+      class Rails40 < Base
+        def match?
+          version >= Gem::Version.new('4.0.0')
+        end
+
+        def adapter
+          ActiveRecord::Base.connection_config[:adapter]
+        end
+      end
+
+      class Rails61 < Base
+        def match?
+          version >= Gem::Version.new('6.1.0')
+        end
+
+        def adapter
+          ActiveRecord::Base.connection_db_config.configuration_hash[:adapter]
+        end
+      end
+
+      def self.name
+        new(NewRelic::Control.instance.env, Gem::Version.new(ActiveRecord::VERSION::STRING)).name
+      end
+
+      attr_reader :env, :version
+
+      def initialize(env, version)
+        @env = env
+        @version = version
+      end
+
+      def name
+        [Rails61, Rails40, Rails30, Base]
```

Comment:
The idea here is that these classes define the boundaries where a breaking change was made in the adapter lookup implementation in ActiveRecord. For example, since there was no change in implementation between Rails 4.0 and Rails 6.0, they can all use the implementation provided by `Rails40`